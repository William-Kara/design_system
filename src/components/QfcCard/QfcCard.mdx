import { ArgTypes, Meta, Story } from '@storybook/blocks';
import QfcCard from '@/components/QfcCard/QfcCard.vue';

<Meta title="Modules/Card" />

# Definition

This is a QfcCard module that is used to display a box with an icon or video animated icon, title, and content.
It's using QfcBox.

Size small and large aren't available yet.

## Props

The following props are available for the QfcCard component:

<ArgTypes of={QfcCard} />

# Usage

To use QfcCard, style needs to be imported separately.

```sass
@import '@website/design-system-vuejs/dist/styles/components/QfcCard.scss';
```

```html
<template>
  <QfcCard size="medium" title="Lorem" content="lorem" icon="timer" src="srcVideo" />
</template>
```

# Declinaison

## FullWidth

<Story name="Primary" id="modules-card--primary" />

## Medium

<Story name="Medium" id="modules-card--medium" />

##### With video icon

<Story name="Video Icon" id="modules-card--video-icon" />
