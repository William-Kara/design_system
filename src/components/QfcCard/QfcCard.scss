.q-card {
  padding: $q-generic-space-9 1.5rem;
  text-align: center;
  font-weight: $light;
  display: flex;
  flex-direction: column;
  gap: $q-gap-l;

  &__icon {
    .material-icons-outlined {
      font-size: $fs-xl;
    }
  }
  &__media {
    mix-blend-mode: lighten;
    max-width: 8rem;
  }
  &__title {
    font-weight: $light;
    font-size: $fs-m-2;
  }
  &__content {
    padding: $q-generic-space-2 1.5rem;
    font-size: $fs-s;
    font-weight: $regular;
    text-align: left;
    overflow: hidden;
  }
}
@include mobile {
  .q-card {
    padding: $q-generic-space-9 1.15rem;
    gap: $q-gap-m;
    &__icon {
      .material-icons-outlined {
        font-size: $fs-l;
      }
    }
    &--is-medium {
      max-width: min(100%, 27.4rem);
    }
  }
}
