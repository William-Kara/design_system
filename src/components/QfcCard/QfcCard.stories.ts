import Card from './QfcCard.vue';
import type { Meta, StoryObj } from '@storybook/vue3';
import videoFile from '../../public/timer-animation.mp4';

const meta: Meta<typeof Card> = {
  component: Card,
  title: 'Modules/Card',
  argTypes: {
    size: {
      type: { name: 'string', required: false },
      defaultValue: 'fullwidth',
    },
    icon: {
      type: { name: 'string', required: true },
      defaultValue: 'visibility',
    },
    srcVideo: {
      type: { name: 'string', required: false },
    },
    title: {
      type: { name: 'string', required: false },
      defaultValue: 'Title',
    },
    content: {
      type: { name: 'string', required: false },
      defaultValue: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Quisquam, aliquam.',
    },
  },
};
export default meta;

type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {
    size: 'fullwidth',
    title: 'Une plateforme accessible à tout moment',
    icon: 'explore',
    content:
      'Grâce à une plateforme accessible partout et à tout moment, vous pourrez retrouver vos formations en cours rapidement et reprendre là où vous en étiez.',
  },
};

export const Medium: Story = {
  args: {
    ...Primary.args,
    size: 'medium',
    icon: 'timer',
  },
};

export const VideoIcon: Story = {
  args: {
    srcVideo: videoFile,
    size: 'medium',
    title: 'Une plateforme accessible à tout moment',
    content:
      'Grâce à une plateforme accessible partout et à tout moment, vous pourrez retrouver vos formations en cours rapidement et reprendre là où vous en étiez.',
  },
};
