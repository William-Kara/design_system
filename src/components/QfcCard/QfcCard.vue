<template>
  <QfcBox
    :class="{ 'q-card': true, 'q-card--is-medium': size === 'medium' }"
    :size="size"
    ref="target"
    v-element-visibility="targetIsVisible"
  >
    <div class="q-card__icon">
      <i v-if="!srcVideo" class="material-icons-outlined">
        {{ icon }}
      </i>
      <video
        v-if="srcVideo && targetIsVisible"
        class="q-card__media"
        :src="srcVideo"
        autoplay
        muted
      />
    </div>

    <h3 class="q-card__title">{{ title }}</h3>

    <p class="q-card__content">
      {{ content }}
    </p>
  </QfcBox>
</template>

<script lang="ts" setup>
import QfcBox from '@/components/QfcBox/QfcBox.vue';
import { useElementVisibility } from '@vueuse/core';
import { ref } from 'vue';

const sizeType = ['small', 'medium', 'large', 'fullwidth'] as const;
type SizeType = (typeof sizeType)[number];

const target = ref(null);

const targetIsVisible = useElementVisibility(target);

interface Props {
  size: SizeType;
  srcVideo: string;
  icon: string;
  title: string;
  content: string;
}

withDefaults(defineProps<Props>(), {
  size: 'fullwidth',
});
</script>

<style lang="scss" scoped src="./QfcCard.scss" />
