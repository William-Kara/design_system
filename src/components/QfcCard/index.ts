import QfcCard from './QfcCard.vue';

export { QfcCard };
export default QfcCard;
