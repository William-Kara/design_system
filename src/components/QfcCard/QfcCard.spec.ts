import { it, describe, beforeEach, expect } from 'vitest';

import { shallowMount } from '@vue/test-utils';
import QfcCard from './QfcCard.vue';

let wrapper: any = null;

const propsValues = {
  size: 'medium',
  icon: 'explore',
  title: 'Title',
  content: 'Content',
};

beforeEach(() => {
  wrapper = shallowMount(QfcCard, {
    propsData: propsValues,
  });
});

describe('QfcCard Component', () => {
  it('should mount properly', () => {
    expect(wrapper).toBeTruthy();
  });

  it('should have correct props', () => {
    expect(wrapper.props()).toEqual(propsValues);
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
