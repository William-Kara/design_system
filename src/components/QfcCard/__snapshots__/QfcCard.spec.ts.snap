// Vitest Snapshot v1

exports[`QfcCard Component > should render correctly 1`] = `"<qfc-box-stub data-v-4d4fb029=\\"\\" size=\\"medium\\" padding=\\"medium\\" class=\\"q-card q-card--is-medium\\"></qfc-box-stub>"`;
